entity-framework/core/testing/unit-testing.md

```diff
@@ -0,0 +1,60 @@
+---
+title: Unit testing with EF Core
+description: Approaches and techniques to doing unit testing with applications using Entity Framework Core
+author: roji
+ms.date: 11/07/2021
+uid: core/testing/unit-testing
+---
+# Unit Testing
+
+In the context of the EF Core applications, unit tests are automated tests which do not involve the database system against which the application runs in production. There are various unit testing approaches which swap the database with a test double, as well as integration tests which do involve the production database system. It's recommend to first read the [testing overview page](xref:core/testing/index) for a comparison of the different testing patterns and guidelines on how to choose the best one for your scenario. This page shows how to use each unit testing approach, and provides code samples for them.
+
+## Repository pattern
+
+The repository pattern is the recommended technique for writing unit tests for EF Core applications - for more background, see [the corresponding section in the overview page](xref:core/testing/index#repository-pattern). The first step of implementing the repository pattern is to extract out your EF Core LINQ queries to a separate layer, which we'll later stub or mock. Here's an example of a repository interface for our blogging system:
+
+[!code-csharp[Main](../../../samples/core/Testing/BusinessLogic/IBloggingRepository.cs?name=IBloggingRepository)]
+
+... and here's a partial sample implementation for production use:
+
+[!code-csharp[Main](../../../samples/core/Testing/BusinessLogic/BloggingRepository.cs?name=BloggingRepository)]
+
+There's not much to it: the repository simply wraps an EF Core context, and exposes methods which execute the database queries and updates on it. A key point to note is that our `GetAllBlogs` method returns `IEnumerable<Blog>`, and not `IQueryable<Blog>`. Returning the latter would mean that query operators can still be composed over the result, requiring EF Core to still involved in translating the query; this would defeat the purpose of having a repository in the first place. `IEnumerable<Blog>` allows us to easily stub or mock what the repository returns.
```

Comment:
```suggestion
There's not much to it: the repository simply wraps an EF Core context, and exposes methods which execute the database queries and updates on it. A key point to note is that our `GetAllBlogs` method returns `IEnumerable<Blog>`, and not `IQueryable<Blog>`. Returning the latter would mean that query operators can still be composed over the result, requiring that EF Core is still involved in translating the query; this would defeat the purpose of having a repository in the first place. `IEnumerable<Blog>` allows us to easily stub or mock what the repository returns.
```